### PR TITLE
Support Postgresql

### LINKED ISSUE #89: Switch from sqlite to postgres
To allow for better performance and scalability for large deployments.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #89

### CODE REVIEW

**Comment by Half-Shot** [comment]
Unit tests won't pass without https://github.com/matrix-org/pipelines/pull/53

File: sygnal.yaml.sample [hunk 1]
```diff
@@ -169,3 +169,10 @@ apps:
   #  # the default is 20.
   #  #max_connections: 20
 
+db:
+  host: localhost
+  database: sygnal
+  user: postgres
```
Comment (anoadragon453) [inline sygnal.yaml.sample]: Surely we shouldn't encourage using the `postgres` user for an application.

File: sygnal/gcmpushkin.py [hunk 2]
```diff
@@ -482,23 +486,14 @@ def __init__(self):
...
-
-        Returns (dict):
-            its canonical registration ID, or `None` if there is no entry.
-        """
-        rows = await self.db.runQuery(
-            "SELECT canonical_reg_id FROM gcm_canonical_reg_id WHERE reg_id = ?",
-            (reg_id,),
+        results = dict(
+            await self.db.runQuery(
+                """SELECT reg_id, canonical_reg_id
```
Comment (anoadragon453) [inline sygnal/gcmpushkin.py]: Can you make this prettier?

File: sygnal/sygnal.py [hunk 3]
```diff
@@ -235,7 +237,7 @@ def check_section(section_name, known_keys, cfgpart=config):
     check_section(
         "access", {"file", "enabled", "x_forwarded_for"}, cfgpart=config["log"]
     )
-    check_section("db", {"dbfile"})
+    check_section("db", {"args"})
```
Comment (anoadragon453) [inline sygnal/sygnal.py]: Shouldn't be `args`.

File: tests/testutils.py [hunk 4]
```diff
@@ -51,13 +95,14 @@ def setUp(self):
         )
 
         while not start_deferred.called:
-            # we need to advance until the pushkins have started up
+            # we need to wait for the database
             self.sygnal.reactor.advance(1)
             self.sygnal.reactor.wait_for_work(lambda: start_deferred.called)
```
Comment (anoadragon453) [inline tests/testutils.py]: It looks like this is still waiting for `start_deferred`, which waits for the pushkins.

File: tests/testutils.py [hunk 5]
```diff
@@ -24,15 +26,56 @@
...
+        conn = psycopg2.connect(
+            database=POSTGRES_DBNAME_FOR_INITIAL_CREATE,
+            user=POSTGRES_USER,
+            password=POSTGRES_PASSWORD,
+            host=POSTGRES_HOST,
+        )
+        conn.autocommit = True
+        cur = conn.cursor()
+        cur.execute("DROP DATABASE IF EXISTS %s;" % (dbname,))
+        cur.execute("CREATE DATABASE %s;" % (dbname,))
```
Comment (anoadragon453) [inline tests/testutils.py]: You need to `conn.commit()` before closing.

File: tests/testutils.py [hunk 6]
```diff
@@ -24,15 +26,56 @@
...
+    def _tear_down_database(self, dbname):
+        conn = psycopg2.connect(
+            database=POSTGRES_DBNAME_FOR_INITIAL_CREATE,
+            user=POSTGRES_USER,
+            password=POSTGRES_PASSWORD,
+            host=POSTGRES_HOST,
+        )
+        conn.autocommit = True
+        cur = conn.cursor()
+        cur.execute("DROP DATABASE %s;" % (dbname,))
```
Comment (anoadragon453) [inline tests/testutils.py]: Same here.

File: sygnal/gcmpushkin.py [hunk 7]
```diff
@@ -465,8 +465,12 @@ def __init__(self):
...
             canonical_reg_id (str): the canonical registration ID for `reg_id`
         """
-        await self.db.runQuery(
-            "INSERT OR REPLACE INTO gcm_canonical_reg_id VALUES (?, ?);",
+        await self.db.runOperation(
+            """
+            INSERT INTO gcm_canonical_reg_id VALUES (%s, %s)
+            ON CONFLICT (reg_id) DO UPDATE
+                 SET canonical_reg_id = EXCLUDED.canonical_reg_id;
+            """,
```
Comment (anoadragon453) [inline sygnal/gcmpushkin.py]: Does `runOperation` `commit()` automatically or do we need to do so explicitly here as well?

File: sygnal/gcmpushkin.py [hunk 8]
```diff
@@ -451,12 +451,12 @@ def __init__(self):
...
 
         Args:
-            db (adbapi.ConnectionPool): database to prepare
+            db (psycopg2.pool.AbstractConnectionPool): database to prepare
 
         """
         self.db = db
 
-        await self.db.runQuery(self.TABLE_CREATE_QUERY)
+        await self.db.runOperation(self.TABLE_CREATE_QUERY)
```
Comment (anoadragon453) [inline sygnal/gcmpushkin.py]: Do we need to `commit()` here?

File: tests/testutils.py (hunk 5 above)
Comment (Half-Shot) [reply tests/testutils.py]: Not in this case. We're using `conn.autocommit` like in https://github.com/matrix-org/synapse/blob/d9239b52571c7e07fa9b035de1ad46bb602ed163/tests/utils.py#L68-L79 because running a DROP and a CREATE inside a single txn isn't allowed.

File: sygnal/gcmpushkin.py (hunk 7 above)
Comment (Half-Shot) [reply sygnal/gcmpushkin.py]: `runOperation` commits automatically

File: sygnal/gcmpushkin.py (hunk 8 above)
Comment (Half-Shot) [reply sygnal/gcmpushkin.py]: same as above.

File: tests/testutils.py (hunk 5 above)
Comment (anoadragon453) [reply tests/testutils.py]: Oh! Missed that, interesting

File: sygnal.yaml.sample [hunk 9]
```diff
@@ -169,3 +169,10 @@ apps:
   #  # the default is 20.
   #  #max_connections: 20
 
+db:
```
Comment (richvdh) [inline sygnal.yaml.sample]: this should probably be nearer the top? and needs some comments. What does it do? what are the defaults if the settings are not given?

File: sygnal/sygnal.py [hunk 10]
```diff
@@ -235,7 +237,6 @@ def check_section(section_name, known_keys, cfgpart=config):
     check_section(
         "access", {"file", "enabled", "x_forwarded_for"}, cfgpart=config["log"]
     )
-    check_section("db", {"dbfile"})
```
Comment (richvdh) [inline sygnal/sygnal.py]: does the updated section not merit any checks?

File: sygnal/gcmpushkin.py [hunk 11]
```diff
@@ -91,7 +91,7 @@ class GcmPushkin(Pushkin):
     Pushkin that relays notifications to Google/Firebase Cloud Messaging.
     """
 
-    UNDERSTOOD_CONFIG_FIELDS = {"type", "api_key"}
+    UNDERSTOOD_CONFIG_FIELDS = {"type", "api_key", "max_connections"}
```
Comment (richvdh) [inline sygnal/gcmpushkin.py]: is this related?

**Comment by anoadragon453** [comment]
We assumed since we assume that nobody else runs this, switching all deployments to postgres wouldn't be an issue.

I know we can't *know* we aren't the only ones running this, but that's what we've assumed in the past.

File: sygnal/sygnal.py (hunk 10 above)
Comment (Half-Shot) [reply sygnal/sygnal.py]: I've added some checks below. It's a bit *ugh* because we are checking for different things depending on the database being used, but it's better than nothing.

File: sygnal.yaml.sample [hunk 12]
```diff
@@ -3,6 +3,21 @@
 # See: matrix.org
 ##
 
+## Database #
```
Comment (richvdh) [inline sygnal.yaml.sample]: could you make this follow the synapse conventions for sample configs? I feel like it's been serving us well there.

**Comment by richvdh** [comment]
> We assumed since we assume that nobody else runs this, switching all deployments to postgres wouldn't be an issue.

A few points that occur to me here. Not to pick on either of you guys, but just some things I think we all need to think about as we try to get ourselves into a more mature situation where we aren't always doing the bare minimum to put out the current fire:

* As you say, we can't assume that we are the only people running this, and the fact that we have done so in the past (and presumably totally messed other people about in the process) isn't a great reason to do so now.
* Even if we *were* the only people running this, I'd be keen that we maintain support for sqlite for a while at least. Operationally, things are much easier if you can separate software upgrades from configuration changes: it means you don't have to change everything at once (and, often more importantly, you don't have to change everything back when it doesn't work.)

Similarly, I'm keen that we document stuff like this clearly, even if we don't expect other people to run it. The reasons for this are:

 * New Vector is now big enough that the people writing the code may or may not be the people deploying it, or - particularly - the people fixing it at 2am. When I've just been woken up to fix a problem, clear documentation is a godsend.
 * Even if we get this deployed and configured and it all runs perfectly, there will come a point in 6 or 12 months when we need to re-deploy it or something. And at that point, we'll have forgotten everything we knew. Rather than have to figure it out again from first principles, we could look at the documentation we wrote while it was fresh in our heads.

In both cases, imagining that you are writing for an external audience will be a great exercise in making sure that we have good documentation, even if that external audience doesn't really exist.

File: sygnal/sygnal.py (hunk 10 above)
Comment (Half-Shot) [reply sygnal/sygnal.py]: I've since added checks that `name` and `args` are present.

File: sygnal.yaml.sample (hunk 12 above)
Comment (Half-Shot) [reply sygnal.yaml.sample]: Have copied across

File: sygnal/gcmpushkin.py [hunk 13]
```diff
@@ -443,7 +443,7 @@ class CanonicalRegIdStore(object):
     def __init__(self):
         self.db = None
 
-    async def setup(self, db):
+    async def setup(self, db, engine):
```
Comment (richvdh) [inline sygnal/gcmpushkin.py]: docstring for `engine` please

File: sygnal/gcmpushkin.py [hunk 14]
```diff
@@ -482,23 +493,15 @@ def __init__(self):
...
-        rows = await self.db.runQuery(
-            "SELECT canonical_reg_id FROM gcm_canonical_reg_id WHERE reg_id = ?",
-            (reg_id,),
+        rows = dict(
+            await self.db.runQuery(
+                """
+            SELECT reg_id, canonical_reg_id
+            FROM gcm_canonical_reg_id
+            WHERE reg_id IN (%s)
+            """
```
Comment (richvdh) [inline sygnal/gcmpushkin.py]: this lot could do with more indentation, for consistency with other `"""` blocks

File: sygnal/gcmpushkin.py [hunk 15]
```diff
@@ -482,23 +493,15 @@ def __init__(self):
             mapping of registration ID to either its canonical registration ID,
             or `None` if there is no entry.
         """
-        return {reg_id: await self.get_canonical_id(reg_id) for reg_id in reg_ids}
```
Comment (richvdh) [inline sygnal/gcmpushkin.py]: not saying it's a bad idea, but... is this actually part of supporting postgres or is it an unrelated change?

File: sygnal/sygnal.py [hunk 16]
```diff
@@ -76,6 +76,17 @@ def __init__(self, config, custom_reactor, tracer=opentracing.tracer):
         observer = twisted_log.PythonLoggingObserver()
         observer.start()
 
+        # Old format db config
+        if config.get("db", None) is not None:
```
Comment (richvdh) [inline sygnal/sygnal.py]: `dict.get` defaults to None, and it's pythonic to skip the explicit `is not None`

```suggestion
        if config.get("db"):
```

File: sygnal/sygnal.py [hunk 17]
```diff
@@ -76,6 +76,17 @@ def __init__(self, config, custom_reactor, tracer=opentracing.tracer):
...
+            logger.warning(
+                "Config includes the legacy 'db' option, please migrate"
+                + "to 'database' instead"
```
Comment (richvdh) [inline sygnal/sygnal.py]: "migrateto" is two words. `+` operator is redundant.

```suggestion
                " to 'database' instead"
```

File: sygnal.yaml.sample [hunk 18]
```diff
@@ -3,6 +3,45 @@
 # See: matrix.org
 ##
 
+# The 'database' setting defines the database that sygnal uses to store all of
```
Comment (richvdh) [inline sygnal.yaml.sample]: this documentation is nice now. :+1:

File: sygnal/gcmpushkin.py (hunk 15 above)
Comment (richvdh) [reply sygnal/gcmpushkin.py]: (for clarity: I'm not saying you need to change it back now, but for future ref, assuming it is unrelated, I'd prefer it to be in a separate PR.)

Comment (Half-Shot) [reply sygnal/gcmpushkin.py]: Yep, I got a bit eager with changing something that seemed like a performance problem. Luckily there are unit tests testing this path. In the future, I will open a separate PR :+1:

File: sygnal/sygnal.py (hunk 16 above)
Comment (Half-Shot) [reply sygnal/sygnal.py]: Good to know :+1:

File: sygnal/gcmpushkin.py [hunk 19]
```diff
@@ -452,11 +452,14 @@ def __init__(self):
 
         Args:
             db (adbapi.ConnectionPool): database to prepare
+            engine (str):
+                Database engine in use. This is either "sqlite" or "postgresql".
```
Comment (richvdh) [inline sygnal/gcmpushkin.py]: it's not in use yet?

```suggestion
                Database engine to use. Shoud be either "sqlite" or "postgresql".
```
